ext/java/org/msgpack/jruby/Packer.java

```diff
@@ -30,20 +35,139 @@ public IRubyObject allocate(Ruby runtime, RubyClass type) {
     }
   }
 
+  static class ExtRegistry {
+    private Ruby runtime;
+    public RubyHash hash;
+    public RubyHash cache;
+
+    public ExtRegistry(Ruby runtime) {
+      this.runtime = runtime;
+      hash = RubyHash.newHash(runtime);
+      cache = RubyHash.newHash(runtime);
+    }
+
+    public ExtRegistry dup() {
+      ExtRegistry copy = new ExtRegistry(this.runtime);
+      copy.hash = (RubyHash) this.hash.dup(runtime.getCurrentContext());
+      copy.cache = RubyHash.newHash(runtime);
+      return copy;
+    }
+
+    public void put(RubyClass klass, int typeId, IRubyObject proc, IRubyObject arg) {
+      RubyArray e = RubyArray.newArray(runtime, new IRubyObject[] { RubyFixnum.newFixnum(runtime, typeId), proc, arg });
```

Comment:
The code in the `lookup` method gets really complicated to follow because you're using arrays as values. Add a `Entry` class with three public instance variables instead. That would mean that the code in `lookup` would use `entry.typeId` instead of `e.entry(0)`.

Again, using `RubyHash` and `RubyArray` gets really messy in Java code. Use them only when you need to return the objects to Ruby code.